// leva/dist/leva.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vectorPlugin = require('./vector-plugin-19a8a46a.cjs.dev.js');
var v8n = require('v8n');
var React = require('react');
require('@radix-ui/react-portal');
var lite = require('dequal/lite');
var tc = require('tinycolor2');
var reactColorful = require('react-colorful');
var shallow = require('zustand/shallow');
var reactUseGesture = require('react-use-gesture');
require('clipboard-polyfill/text');
require('@radix-ui/react-tooltip');
var reactDropzone = require('react-dropzone');
var create = require('zustand');
var ReactDOM = require('react-dom');
var merge = require('merge-value');
require('@stitches/react');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var v8n__default = /*#__PURE__*/_interopDefault(v8n);
var React__default = /*#__PURE__*/_interopDefault(React);
var tc__default = /*#__PURE__*/_interopDefault(tc);
var shallow__default = /*#__PURE__*/_interopDefault(shallow);
var create__default = /*#__PURE__*/_interopDefault(create);
var ReactDOM__default = /*#__PURE__*/_interopDefault(ReactDOM);
var merge__default = /*#__PURE__*/_interopDefault(merge);

const join = (...args) => args.filter(Boolean).join('.');
function getKeyPath(path) {
  const dir = path.split('.');
  return [dir.pop(), dir.join('.') || undefined];
}

function getValuesForPaths(data, paths) {
  return Object.entries(vectorPlugin.pick(data, paths)).reduce((acc, [path, {
    value,
    disabled,
    key
  }]) => {
    acc[key] = disabled ? undefined : value;
    return acc;
  }, {});
}

function useCompareMemoize(value, deep) {
  const ref = React.useRef();
  const compare = deep ? lite.dequal : shallow__default['default'];

  if (!compare(value, ref.current)) {
    ref.current = value;
  }

  return ref.current;
}

function useDeepMemo(fn, deps) {
  return React.useMemo(fn, useCompareMemoize(deps, true));
}

function useToggle(toggled) {
  const wrapperRef = React.useRef(null);
  const contentRef = React.useRef(null);
  const firstRender = React.useRef(true);
  React.useLayoutEffect(() => {
    if (!toggled) {
      wrapperRef.current.style.height = '0px';
      wrapperRef.current.style.overflow = 'hidden';
    }
  }, []);
  React.useEffect(() => {
    if (firstRender.current) {
      firstRender.current = false;
      return;
    }

    let timeout;
    const ref = wrapperRef.current;

    const fixHeight = () => {
      if (toggled) {
        ref.style.removeProperty('height');
        ref.style.removeProperty('overflow');
        contentRef.current.scrollIntoView({
          behavior: 'smooth',
          block: 'nearest'
        });
      }
    };

    ref.addEventListener('transitionend', fixHeight, {
      once: true
    });
    const {
      height
    } = contentRef.current.getBoundingClientRect();
    ref.style.height = height + 'px';

    if (!toggled) {
      ref.style.overflow = 'hidden';
      timeout = window.setTimeout(() => ref.style.height = '0px', 50);
    }

    return () => {
      ref.removeEventListener('transitionend', fixHeight);
      clearTimeout(timeout);
    };
  }, [toggled]);
  return {
    wrapperRef,
    contentRef
  };
}

const useVisiblePaths = store => {
  const [paths, setPaths] = React.useState(store.getVisiblePaths());
  React.useEffect(() => {
    setPaths(store.getVisiblePaths());
    const unsub = store.useStore.subscribe(setPaths, store.getVisiblePaths, shallow__default['default']);
    return () => unsub();
  }, [store]);
  return paths;
};

function useValuesForPath(store, paths, initialData) {
  const init = React.useRef(true);
  const valuesForPath = store.useStore(s => {
    const data = init.current ? initialData : s.data;
    return getValuesForPaths(data, paths);
  }, shallow__default['default']);
  init.current = false;
  return valuesForPath;
}

function usePopin(margin = 3) {
  const popinRef = React.useRef(null);
  const wrapperRef = React.useRef(null);
  const [shown, setShow] = React.useState(false);
  const show = React.useCallback(() => setShow(true), []);
  const hide = React.useCallback(() => setShow(false), []);
  React.useLayoutEffect(() => {
    if (shown) {
      const {
        bottom,
        top,
        left
      } = popinRef.current.getBoundingClientRect();
      const {
        height
      } = wrapperRef.current.getBoundingClientRect();
      const direction = bottom + height > window.innerHeight - 40 ? 'up' : 'down';
      wrapperRef.current.style.position = 'fixed';
      wrapperRef.current.style.zIndex = '10000';
      wrapperRef.current.style.left = left + 'px';
      if (direction === 'down') wrapperRef.current.style.top = bottom + margin + 'px';else wrapperRef.current.style.bottom = window.innerHeight - top + margin + 'px';
    }
  }, [margin, shown]);
  return {
    popinRef,
    wrapperRef,
    shown,
    show,
    hide
  };
}

const schema$5 = (_o, s) => v8n__default['default']().schema({
  options: v8n__default['default']().passesAnyOf(v8n__default['default']().object(), v8n__default['default']().array())
}).test(s);
const normalize$4 = input => {
  let {
    value,
    options
  } = input;
  let keys;
  let values;

  if (Array.isArray(options)) {
    values = options;
    keys = options.map(String);
  } else {
    values = Object.values(options);
    keys = Object.keys(options);
  }

  if (!('value' in input)) value = values[0];else if (!values.includes(value)) {
    keys.unshift(String(value));
    values.unshift(value);
  }
  if (!Object.values(options).includes(value)) options[String(value)] = value;
  return {
    value,
    settings: {
      keys,
      values
    }
  };
};

var props$5 = /*#__PURE__*/Object.freeze({
  __proto__: null,
  schema: schema$5,
  normalize: normalize$4
});

const SelectContainer = vectorPlugin.styled('div', {
  $leva__flexCenter: '',
  position: 'relative',
  '> svg': {
    pointerEvents: 'none',
    position: 'absolute',
    right: '$leva__md'
  }
});
const NativeSelect = vectorPlugin.styled('select', {
  position: 'absolute',
  top: 0,
  left: 0,
  width: '100%',
  height: '100%',
  opacity: 0
});
const PresentationalSelect = vectorPlugin.styled('div', {
  display: 'flex',
  alignItems: 'center',
  width: '100%',
  height: '$leva__rowHeight',
  backgroundColor: '$leva__elevation3',
  borderRadius: '$leva__sm',
  padding: '0 $leva__sm',
  cursor: 'pointer',
  [`${NativeSelect}:focus + &`]: {
    $leva__focusStyle: ''
  },
  [`${NativeSelect}:hover + &`]: {
    $leva__hoverStyle: ''
  }
});

function Select({
  displayValue,
  onUpdate,
  id,
  settings
}) {
  const {
    keys
  } = settings;
  return React__default['default'].createElement(SelectContainer, null, React__default['default'].createElement(NativeSelect, {
    id: id,
    value: displayValue,
    onChange: e => onUpdate(e.currentTarget.value)
  }, keys.map(key => React__default['default'].createElement("option", {
    key: key,
    value: key
  }, key))), React__default['default'].createElement(PresentationalSelect, null, displayValue), React__default['default'].createElement(vectorPlugin.Chevron, {
    toggled: true
  }));
}

function SelectComponent() {
  const {
    label,
    displayValue,
    onUpdate,
    id,
    settings
  } = vectorPlugin.useInputContext();
  return React__default['default'].createElement(vectorPlugin.Row, {
    input: true
  }, React__default['default'].createElement(vectorPlugin.Label, null, label), React__default['default'].createElement(Select, {
    id: id,
    displayValue: displayValue,
    onUpdate: onUpdate,
    settings: settings
  }));
}

var select = vectorPlugin.createInternalPlugin(vectorPlugin._objectSpread2({
  component: SelectComponent
}, props$5));

const convertMap = {
  rgb: 'toRgb',
  hsl: 'toHsl',
  hsv: 'toHsv',
  hex: 'toHexString',
  hex8: 'toHex8String'
};
v8n__default['default'].extend({
  color: () => value => tc__default['default'](value).isValid()
});
const schema$4 = o => v8n__default['default']().color().test(o);

function convert(color, {
  format,
  hasAlpha,
  isString
}) {
  const _format = format === 'hex' && hasAlpha ? 'hex8' : format;

  if (isString) return color.toString(_format);

  const colorObj = color[convertMap[_format]]();

  return hasAlpha ? colorObj : vectorPlugin.omit(colorObj, ['a']);
}

const sanitize$4 = (v, settings) => {
  const color = tc__default['default'](v);
  if (!color.isValid()) throw Error('Invalid color');
  return convert(color, settings);
};
const format$1 = (v, settings) => {
  return convert(tc__default['default'](v), vectorPlugin._objectSpread2(vectorPlugin._objectSpread2({}, settings), {}, {
    isString: true,
    format: 'hex'
  }));
};
const normalize$3 = ({
  value
}) => {
  const color = tc__default['default'](value);

  const _f = color.getFormat();

  const format = _f === 'name' || _f === 'hex8' ? 'hex' : _f;
  const hasAlpha = typeof value === 'object' ? 'a' in value : _f === 'hex8' || /^(rgba)|(hsla)|(hsva)/.test(value);
  const settings = {
    format,
    hasAlpha,
    isString: typeof value === 'string'
  };
  return {
    value: sanitize$4(value, settings),
    settings
  };
};

var props$4 = /*#__PURE__*/Object.freeze({
  __proto__: null,
  schema: schema$4,
  sanitize: sanitize$4,
  format: format$1,
  normalize: normalize$3
});

const ColorPreview = vectorPlugin.styled('div', {
  position: 'relative',
  boxSizing: 'border-box',
  borderRadius: '$leva__sm',
  overflow: 'hidden',
  cursor: 'pointer',
  height: '$leva__rowHeight',
  width: '$leva__rowHeight',
  backgroundColor: '#fff',
  backgroundImage: `url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill-opacity=".05"><path d="M8 0h8v8H8zM0 8h8v8H0z"/></svg>')`,
  $leva__inputStyle: '',
  $leva__hover: '',
  zIndex: 1,
  variants: {
    active: {
      true: {
        $leva__inputStyle: '$leva__accent1'
      }
    }
  },
  '&::before': {
    content: '""',
    position: 'absolute',
    top: 0,
    bottom: 0,
    right: 0,
    left: 0,
    backgroundColor: 'currentColor',
    zIndex: 1
  }
});
const PickerContainer = vectorPlugin.styled('div', {
  position: 'relative',
  display: 'grid',
  gridTemplateColumns: '$sizes$leva__rowHeight auto',
  columnGap: '$leva__colGap',
  alignItems: 'center'
});
const PickerWrapper = vectorPlugin.styled('div', {
  width: '$leva__colorPickerWidth',
  height: '$leva__colorPickerHeight',
  '.react-colorful': {
    width: '100%',
    height: '100%',
    boxShadow: '$leva__level2',
    cursor: 'crosshair'
  },
  '.react-colorful__saturation': {
    borderRadius: '$leva__sm $leva__sm 0 0'
  },
  '.react-colorful__alpha, .react-colorful__hue': {
    height: 10
  },
  '.react-colorful__last-control': {
    borderRadius: '0 0 $leva__sm $leva__sm'
  },
  '.react-colorful__pointer': {
    height: 12,
    width: 12
  }
});

function convertToRgb(value, format) {
  return format !== 'rgb' ? tc__default['default'](value).toRgb() : value;
}

function Color({
  value,
  displayValue,
  settings,
  onUpdate
}) {
  const {
    format,
    hasAlpha
  } = settings;
  const {
    popinRef,
    wrapperRef,
    shown,
    show,
    hide
  } = usePopin();
  const timer = React.useRef(0);
  const [initialRgb, setInitialRgb] = React.useState(() => convertToRgb(value, format));
  const ColorPicker = hasAlpha ? reactColorful.RgbaColorPicker : reactColorful.RgbColorPicker;

  const showPicker = () => {
    setInitialRgb(convertToRgb(value, format));
    show();
  };

  const hideAfterDelay = () => {
    timer.current = window.setTimeout(hide, 500);
  };

  return React__default['default'].createElement(React__default['default'].Fragment, null, React__default['default'].createElement(ColorPreview, {
    ref: popinRef,
    active: shown,
    onClick: () => showPicker(),
    style: {
      color: displayValue
    }
  }), shown && React__default['default'].createElement(vectorPlugin.Portal, null, React__default['default'].createElement(vectorPlugin.Overlay, {
    onPointerUp: hide
  }), React__default['default'].createElement(PickerWrapper, {
    ref: wrapperRef,
    onMouseEnter: () => window.clearTimeout(timer.current),
    onMouseLeave: e => e.buttons === 0 && hideAfterDelay()
  }, React__default['default'].createElement(ColorPicker, {
    color: initialRgb,
    onChange: onUpdate
  }))));
}
function ColorComponent() {
  const {
    value,
    displayValue,
    label,
    onChange,
    onUpdate,
    settings
  } = vectorPlugin.useInputContext();
  return React__default['default'].createElement(vectorPlugin.Row, {
    input: true
  }, React__default['default'].createElement(vectorPlugin.Label, null, label), React__default['default'].createElement(PickerContainer, null, React__default['default'].createElement(Color, {
    value: value,
    displayValue: displayValue,
    onChange: onChange,
    onUpdate: onUpdate,
    settings: settings
  }), React__default['default'].createElement(vectorPlugin.ValueInput, {
    value: displayValue,
    onChange: onChange,
    onUpdate: onUpdate
  })));
}

var color = vectorPlugin.createInternalPlugin(vectorPlugin._objectSpread2({
  component: ColorComponent
}, props$4));

const schema$3 = o => v8n__default['default']().string().test(o);
const sanitize$3 = v => {
  if (typeof v !== 'string') throw Error(`Invalid string`);
  return v;
};

var props$3 = /*#__PURE__*/Object.freeze({
  __proto__: null,
  schema: schema$3,
  sanitize: sanitize$3
});

function String$1({
  displayValue,
  onUpdate,
  onChange
}) {
  return React__default['default'].createElement(vectorPlugin.ValueInput, {
    value: displayValue,
    onUpdate: onUpdate,
    onChange: onChange
  });
}
function StringComponent() {
  const {
    label,
    displayValue,
    onUpdate,
    onChange
  } = vectorPlugin.useInputContext();
  return React__default['default'].createElement(vectorPlugin.Row, {
    input: true
  }, React__default['default'].createElement(vectorPlugin.Label, null, label), React__default['default'].createElement(String$1, {
    displayValue: displayValue,
    onUpdate: onUpdate,
    onChange: onChange
  }));
}

var string = vectorPlugin.createInternalPlugin(vectorPlugin._objectSpread2({
  component: StringComponent
}, props$3));

const schema$2 = o => v8n__default['default']().boolean().test(o);
const sanitize$2 = v => {
  if (typeof v !== 'boolean') throw Error('Invalid boolean');
  return v;
};

var props$2 = /*#__PURE__*/Object.freeze({
  __proto__: null,
  schema: schema$2,
  sanitize: sanitize$2
});

const StyledInputWrapper = vectorPlugin.styled('div', {
  position: 'relative',
  $leva__flex: '',
  height: '$leva__rowHeight',
  input: {
    $leva__reset: '',
    height: 0,
    width: 0,
    opacity: 0,
    margin: 0
  },
  label: {
    position: 'relative',
    $leva__flexCenter: '',
    WebkitUserSelect: 'none',
    userSelect: 'none',
    cursor: 'pointer',
    height: '$leva__checkboxSize',
    width: '$leva__checkboxSize',
    backgroundColor: '$leva__elevation3',
    borderRadius: '$leva__sm',
    $leva__hover: ''
  },
  'input:focus + label': {
    $leva__focusStyle: ''
  },
  'input:focus:checked + label, input:checked + label:hover': {
    $leva__hoverStyle: '$leva__accent3'
  },
  'input + label:active': {
    backgroundColor: '$leva__accent1'
  },
  'input:checked + label:active': {
    backgroundColor: '$leva__accent1'
  },
  'label > svg': {
    display: 'none',
    width: '90%',
    height: '90%',
    stroke: '$leva__highlight3'
  },
  'input:checked + label': {
    backgroundColor: '$leva__accent2'
  },
  'input:checked + label > svg': {
    display: 'block'
  }
});

function Boolean$1({
  value,
  onUpdate,
  id
}) {
  return React__default['default'].createElement(StyledInputWrapper, null, React__default['default'].createElement("input", {
    id: id,
    type: "checkbox",
    checked: value,
    onChange: e => onUpdate(e.currentTarget.checked)
  }), React__default['default'].createElement("label", {
    htmlFor: id
  }, React__default['default'].createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 24 24"
  }, React__default['default'].createElement("path", {
    strokeLinecap: "round",
    strokeLinejoin: "round",
    strokeWidth: 2,
    d: "M5 13l4 4L19 7"
  }))));
}
function BooleanComponent() {
  const {
    label,
    value,
    onUpdate,
    id
  } = vectorPlugin.useInputContext();
  return React__default['default'].createElement(vectorPlugin.Row, {
    input: true
  }, React__default['default'].createElement(vectorPlugin.Label, null, label), React__default['default'].createElement(Boolean$1, {
    value: value,
    onUpdate: onUpdate,
    id: id
  }));
}

var boolean = vectorPlugin.createInternalPlugin(vectorPlugin._objectSpread2({
  component: BooleanComponent
}, props$2));

function Vector3dComponent() {
  const {
    label,
    displayValue,
    onUpdate,
    settings
  } = vectorPlugin.useInputContext();
  return React__default['default'].createElement(vectorPlugin.Row, {
    input: true
  }, React__default['default'].createElement(vectorPlugin.Label, null, label), React__default['default'].createElement(vectorPlugin.Vector, {
    value: displayValue,
    settings: settings,
    onUpdate: onUpdate
  }));
}

var vector3d = vectorPlugin.createInternalPlugin(vectorPlugin._objectSpread2({
  component: Vector3dComponent
}, vectorPlugin.getVectorPlugin(['x', 'y', 'z'])));

const JoystickTrigger = vectorPlugin.styled('div', {
  $leva__flexCenter: '',
  position: 'relative',
  backgroundColor: '$leva__elevation3',
  borderRadius: '$leva__sm',
  cursor: 'pointer',
  height: '$leva__rowHeight',
  width: '$leva__rowHeight',
  $leva__draggable: '',
  $leva__hover: '',
  '&:active': {
    cursor: 'none'
  },
  '&::after': {
    content: '""',
    backgroundColor: '$leva__accent2',
    height: 4,
    width: 4,
    borderRadius: 2
  }
});
const JoystickPlayground = vectorPlugin.styled('div', {
  $leva__flexCenter: '',
  width: '$leva__joystickWidth',
  height: '$leva__joystickHeight',
  borderRadius: '$leva__sm',
  boxShadow: '$leva__level2',
  position: 'fixed',
  zIndex: 10000,
  overflow: 'hidden',
  $leva__draggable: '',
  transform: 'translate(-50%, -50%)',
  variants: {
    isOutOfBounds: {
      true: {
        backgroundColor: '$leva__elevation1'
      },
      false: {
        backgroundColor: '$leva__elevation3'
      }
    }
  },
  '> div': {
    position: 'absolute',
    $leva__flexCenter: '',
    borderStyle: 'solid',
    borderWidth: 1,
    borderColor: '$leva__highlight1',
    backgroundColor: '$leva__elevation3',
    width: '80%',
    height: '80%',
    '&::after,&::before': {
      content: '""',
      position: 'absolute',
      zindex: 10,
      backgroundColor: '$leva__highlight1'
    },
    '&::before': {
      width: '100%',
      height: 1
    },
    '&::after': {
      height: '100%',
      width: 1
    }
  },
  '> span': {
    position: 'relative',
    zindex: 100,
    width: 10,
    height: 10,
    backgroundColor: '$leva__accent2',
    borderRadius: '50%'
  }
});

function Joystick({
  value,
  settings,
  onUpdate
}) {
  const timeout = React.useRef();
  const outOfBoundsX = React.useRef(0);
  const outOfBoundsY = React.useRef(0);
  const stepMultiplier = React.useRef(1);
  const [joystickShown, setShowJoystick] = React.useState(false);
  const [isOutOfBounds, setIsOutOfBounds] = React.useState(false);
  const [spanRef, set] = vectorPlugin.useTransform();
  const joystickeRef = React.useRef(null);
  const playgroundRef = React.useRef(null);
  React.useLayoutEffect(() => {
    if (joystickShown) {
      const {
        top,
        left,
        width,
        height
      } = joystickeRef.current.getBoundingClientRect();
      playgroundRef.current.style.left = left + width / 2 + 'px';
      playgroundRef.current.style.top = top + height / 2 + 'px';
    }
  }, [joystickShown]);
  const {
    keys: [v1, v2]
  } = settings;
  const {
    [v1]: {
      step: stepV1
    },
    [v2]: {
      step: stepV2
    }
  } = settings;
  const wpx = vectorPlugin.useTh('sizes', 'leva__joystickWidth');
  const hpx = vectorPlugin.useTh('sizes', 'leva__joystickHeight');
  const w = parseFloat(wpx) * 0.8 / 2;
  const h = parseFloat(hpx) * 0.8 / 2;
  const startOutOfBounds = React.useCallback(() => {
    if (timeout.current) return;
    setIsOutOfBounds(true);
    if (outOfBoundsX.current) set({
      x: outOfBoundsX.current * w
    });
    if (outOfBoundsY.current) set({
      y: outOfBoundsY.current * -h
    });
    timeout.current = window.setInterval(() => {
      onUpdate(v => {
        const incX = stepV1 * outOfBoundsX.current * stepMultiplier.current;
        const incY = stepV2 * outOfBoundsY.current * stepMultiplier.current;
        return Array.isArray(v) ? {
          [v1]: v[0] + incX,
          [v2]: v[1] + incY
        } : {
          [v1]: v[v1] + incX,
          [v2]: v[v2] + incY
        };
      });
    }, 16);
  }, [w, h, onUpdate, set, stepV1, stepV2, v1, v2]);
  const endOutOfBounds = React.useCallback(() => {
    window.clearTimeout(timeout.current);
    timeout.current = undefined;
    setIsOutOfBounds(false);
  }, []);
  React.useEffect(() => {
    function setStepMultiplier(event) {
      stepMultiplier.current = vectorPlugin.multiplyStep(event);
    }

    window.addEventListener('keydown', setStepMultiplier);
    window.addEventListener('keyup', setStepMultiplier);
    return () => {
      window.clearTimeout(timeout.current);
      window.removeEventListener('keydown', setStepMultiplier);
      window.removeEventListener('keyup', setStepMultiplier);
    };
  }, []);
  const bind = reactUseGesture.useDrag(({
    first,
    active,
    delta: [dx, dy],
    movement: [mx, my]
  }) => {
    if (first) setShowJoystick(true);

    const _x = vectorPlugin.clamp(mx, -w, w);

    const _y = vectorPlugin.clamp(my, -h, h);

    outOfBoundsX.current = Math.abs(mx) > Math.abs(_x) ? Math.sign(mx - _x) : 0;
    outOfBoundsY.current = Math.abs(my) > Math.abs(_y) ? Math.sign(_y - my) : 0;
    let newX = value[v1];
    let newY = value[v2];

    if (active) {
      if (!outOfBoundsX.current) {
        newX += dx * stepV1 * stepMultiplier.current;
        set({
          x: _x
        });
      }

      if (!outOfBoundsY.current) {
        newY -= dy * stepV2 * stepMultiplier.current;
        set({
          y: _y
        });
      }

      if (outOfBoundsX.current || outOfBoundsY.current) startOutOfBounds();else endOutOfBounds();
      onUpdate({
        [v1]: newX,
        [v2]: newY
      });
    } else {
      setShowJoystick(false);
      outOfBoundsX.current = 0;
      outOfBoundsY.current = 0;
      set({
        x: 0,
        y: 0
      });
      endOutOfBounds();
    }
  });
  return React__default['default'].createElement(JoystickTrigger, vectorPlugin._extends({
    ref: joystickeRef
  }, bind()), joystickShown && React__default['default'].createElement(vectorPlugin.Portal, null, React__default['default'].createElement(JoystickPlayground, {
    ref: playgroundRef,
    isOutOfBounds: isOutOfBounds
  }, React__default['default'].createElement("div", null), React__default['default'].createElement("span", {
    ref: spanRef
  }))));
}

const Container$1 = vectorPlugin.styled('div', {
  display: 'grid',
  columnGap: '$leva__colGap',
  variants: {
    withJoystick: {
      true: {
        gridTemplateColumns: '$sizes$leva__rowHeight auto'
      },
      false: {
        gridTemplateColumns: 'auto'
      }
    }
  }
});
function Vector2dComponent() {
  const {
    label,
    displayValue,
    onUpdate,
    settings
  } = vectorPlugin.useInputContext();
  return React__default['default'].createElement(vectorPlugin.Row, {
    input: true
  }, React__default['default'].createElement(vectorPlugin.Label, null, label), React__default['default'].createElement(Container$1, {
    withJoystick: settings.joystick
  }, settings.joystick && React__default['default'].createElement(Joystick, {
    value: displayValue,
    settings: settings,
    onUpdate: onUpdate
  }), React__default['default'].createElement(vectorPlugin.Vector, {
    value: displayValue,
    settings: settings,
    onUpdate: onUpdate
  })));
}

const plugin = vectorPlugin.getVectorPlugin(['x', 'y']);

const normalize$2 = (_ref) => {
  let {
    joystick = true
  } = _ref,
      input = vectorPlugin._objectWithoutProperties(_ref, ["joystick"]);

  const {
    value,
    settings
  } = plugin.normalize(input);
  return {
    value,
    settings: vectorPlugin._objectSpread2(vectorPlugin._objectSpread2({}, settings), {}, {
      joystick
    })
  };
};

var vector2d = vectorPlugin.createInternalPlugin(vectorPlugin._objectSpread2(vectorPlugin._objectSpread2({
  component: Vector2dComponent
}, plugin), {}, {
  normalize: normalize$2
}));

const sanitize$1 = v => {
  if (v === undefined) return undefined;

  if (v instanceof File) {
    try {
      return URL.createObjectURL(v);
    } catch (e) {
      return undefined;
    }
  }

  if (typeof v === 'string' && v.indexOf('blob:') === 0) return v;
  throw Error(`Invalid image format [undefined | blob | File].`);
};
const schema$1 = (_o, s) => typeof s === 'object' && 'image' in s;
const normalize$1 = ({
  image
}) => {
  return {
    value: image
  };
};

var props$1 = /*#__PURE__*/Object.freeze({
  __proto__: null,
  sanitize: sanitize$1,
  schema: schema$1,
  normalize: normalize$1
});

const ImageContainer = vectorPlugin.styled('div', {
  position: 'relative',
  display: 'grid',
  gridTemplateColumns: '$sizes$leva__rowHeight auto 20px',
  columnGap: '$leva__colGap',
  alignItems: 'center'
});
const DropZone = vectorPlugin.styled('div', {
  $leva__flexCenter: '',
  overflow: 'hidden',
  height: '$leva__rowHeight',
  background: '$leva__elevation3',
  textAlign: 'center',
  color: 'inherit',
  borderRadius: '$leva__sm',
  outline: 'none',
  WebkitUserSelect: 'none',
  userSelect: 'none',
  cursor: 'pointer',
  $leva__inputStyle: '',
  $leva__hover: '',
  $leva__focusWithin: '',
  $leva__active: '$leva__accent1 $leva__elevation1',
  variants: {
    isDragAccept: {
      true: {
        $leva__inputStyle: '$leva__accent1',
        backgroundColor: '$leva__elevation1'
      }
    }
  }
});
const ImagePreview = vectorPlugin.styled('div', {
  boxSizing: 'border-box',
  borderRadius: '$leva__sm',
  height: '$leva__rowHeight',
  width: '$leva__rowHeight',
  $leva__inputStyle: '',
  backgroundSize: 'cover',
  backgroundPosition: 'center',
  variants: {
    hasImage: {
      true: {
        cursor: 'pointer',
        $leva__hover: '',
        $leva__active: ''
      }
    }
  }
});
const ImageLargePreview = vectorPlugin.styled('div', {
  $leva__flexCenter: '',
  width: '$leva__imagePreviewWidth',
  height: '$leva__imagePreviewHeight',
  borderRadius: '$leva__sm',
  boxShadow: '$leva__level2',
  pointerEvents: 'none',
  $leva__inputStyle: '',
  backgroundSize: 'cover',
  backgroundPosition: 'center'
});
const Instructions = vectorPlugin.styled('div', {
  fontSize: '0.8em',
  height: '100%',
  padding: '$leva__rowGap $leva__md'
});
const Remove = vectorPlugin.styled('div', {
  $leva__flexCenter: '',
  top: '0',
  right: '0',
  marginRight: '$leva__sm',
  height: '100%',
  cursor: 'pointer',
  variants: {
    disabled: {
      true: {
        color: '$leva__elevation3',
        cursor: 'default'
      }
    }
  },
  '&::after,&::before': {
    content: '""',
    position: 'absolute',
    height: 2,
    width: 10,
    borderRadius: 1,
    backgroundColor: 'currentColor'
  },
  '&::after': {
    transform: 'rotate(45deg)'
  },
  '&::before': {
    transform: 'rotate(-45deg)'
  }
});

function ImageComponent() {
  const {
    label,
    value,
    onUpdate
  } = vectorPlugin.useInputContext();
  const {
    popinRef,
    wrapperRef,
    shown,
    show,
    hide
  } = usePopin();
  const onDrop = React.useCallback(acceptedFiles => {
    if (acceptedFiles.length) onUpdate(acceptedFiles[0]);
  }, [onUpdate]);
  const clear = React.useCallback(e => {
    e.stopPropagation();
    onUpdate(undefined);
  }, [onUpdate]);
  const {
    getRootProps,
    getInputProps,
    isDragAccept
  } = reactDropzone.useDropzone({
    maxFiles: 1,
    accept: 'image/*',
    onDrop
  });
  return React__default['default'].createElement(vectorPlugin.Row, {
    input: true
  }, React__default['default'].createElement(vectorPlugin.Label, null, label), React__default['default'].createElement(ImageContainer, null, React__default['default'].createElement(ImagePreview, {
    ref: popinRef,
    hasImage: !!value,
    onPointerDown: () => !!value && show(),
    onPointerUp: hide,
    style: {
      backgroundImage: value ? `url(${value})` : 'none'
    }
  }), shown && !!value && React__default['default'].createElement(vectorPlugin.Portal, null, React__default['default'].createElement(vectorPlugin.Overlay, {
    onPointerUp: hide,
    style: {
      cursor: 'pointer'
    }
  }), React__default['default'].createElement(ImageLargePreview, {
    ref: wrapperRef,
    style: {
      backgroundImage: `url(${value})`
    }
  })), React__default['default'].createElement(DropZone, getRootProps({
    isDragAccept
  }), React__default['default'].createElement("input", getInputProps()), React__default['default'].createElement(Instructions, null, isDragAccept ? 'drop image' : 'click or drop')), React__default['default'].createElement(Remove, {
    onClick: clear,
    disabled: !value
  })));
}

var image = vectorPlugin.createInternalPlugin(vectorPlugin._objectSpread2({
  component: ImageComponent
}, props$1));

const number = v8n__default['default']().number();
const schema = (o, s) => v8n__default['default']().array().length(2).every.number().test(o) && v8n__default['default']().schema({
  min: number,
  max: number
}).test(s);
const format = v => ({
  min: v[0],
  max: v[1]
});
const sanitize = (value, {
  bounds: [MIN, MAX]
}, prevValue) => {
  const _newValue = {
    min: prevValue[0],
    max: prevValue[1]
  };

  const {
    min,
    max
  } = vectorPlugin._objectSpread2(vectorPlugin._objectSpread2({}, _newValue), value);

  return [vectorPlugin.clamp(Number(min), MIN, Math.max(MIN, max)), vectorPlugin.clamp(Number(max), Math.min(MAX, min), MAX)];
};
const normalize = ({
  value,
  min,
  max
}) => {
  const boundsSettings = {
    min,
    max
  };
  const settings = vectorPlugin.normalizeKeyedNumberSettings(format(value), {
    min: boundsSettings,
    max: boundsSettings
  });
  const bounds = [min, max];
  return {
    value,
    settings: vectorPlugin._objectSpread2(vectorPlugin._objectSpread2({}, settings), {}, {
      bounds
    })
  };
};

var props = /*#__PURE__*/Object.freeze({
  __proto__: null,
  schema: schema,
  format: format,
  sanitize: sanitize,
  normalize: normalize
});

const Container = vectorPlugin.styled('div', {
  display: 'grid',
  columnGap: '$leva__colGap',
  gridTemplateColumns: 'auto calc($sizes$leva__numberInputMinWidth * 2 + $space$leva__rowGap)'
});

function IntervalSlider(_ref) {
  let {
    value,
    bounds: [min, max],
    onDrag
  } = _ref,
      settings = vectorPlugin._objectWithoutProperties(_ref, ["value", "bounds", "onDrag"]);

  const ref = React.useRef(null);
  const minScrubberRef = React.useRef(null);
  const maxScrubberRef = React.useRef(null);
  const rangeWidth = React.useRef(0);
  const scrubberWidth = vectorPlugin.useTh('sizes', 'leva__scrubberWidth');
  const bind = vectorPlugin.useDrag(({
    event,
    first,
    xy: [x],
    movement: [mx],
    memo: _memo = {}
  }) => {
    if (first) {
      const {
        width,
        left
      } = ref.current.getBoundingClientRect();
      rangeWidth.current = width - parseFloat(scrubberWidth);
      const targetIsScrub = (event === null || event === void 0 ? void 0 : event.target) === minScrubberRef.current || (event === null || event === void 0 ? void 0 : event.target) === maxScrubberRef.current;
      _memo.pos = vectorPlugin.invertedRange((x - left) / width, min, max);
      const delta = Math.abs(_memo.pos - value.min) - Math.abs(_memo.pos - value.max);
      _memo.key = delta < 0 || delta === 0 && _memo.pos <= value.min ? 'min' : 'max';
      if (targetIsScrub) _memo.pos = value[_memo.key];
    }

    const newValue = _memo.pos + vectorPlugin.invertedRange(mx / rangeWidth.current, 0, max - min);
    onDrag({
      [_memo.key]: vectorPlugin.sanitizeStep(newValue, settings[_memo.key])
    });
    return _memo;
  });
  const minStyle = `calc(${vectorPlugin.range(value.min, min, max)} * (100% - ${scrubberWidth}))`;
  const maxStyle = `calc(${1 - vectorPlugin.range(value.max, min, max)} * (100% - ${scrubberWidth}))`;
  return React__default['default'].createElement(vectorPlugin.RangeWrapper, vectorPlugin._extends({
    ref: ref
  }, bind()), React__default['default'].createElement(vectorPlugin.Range, null, React__default['default'].createElement(vectorPlugin.Indicator, {
    style: {
      left: minStyle,
      right: maxStyle
    }
  })), React__default['default'].createElement(vectorPlugin.Scrubber, {
    position: "left",
    ref: minScrubberRef,
    style: {
      left: minStyle
    }
  }), React__default['default'].createElement(vectorPlugin.Scrubber, {
    position: "right",
    ref: maxScrubberRef,
    style: {
      right: maxStyle
    }
  }));
}

function IntervalComponent() {
  const {
    label,
    displayValue,
    onUpdate,
    settings
  } = vectorPlugin.useInputContext();

  const _settings = vectorPlugin._objectWithoutProperties(settings, ["bounds"]);

  return React__default['default'].createElement(React__default['default'].Fragment, null, React__default['default'].createElement(vectorPlugin.Row, {
    input: true
  }, React__default['default'].createElement(vectorPlugin.Label, null, label), React__default['default'].createElement(Container, null, React__default['default'].createElement(IntervalSlider, vectorPlugin._extends({
    value: displayValue
  }, settings, {
    onDrag: onUpdate
  })), React__default['default'].createElement(vectorPlugin.Vector, {
    value: displayValue,
    settings: _settings,
    onUpdate: onUpdate,
    hideNumberLabels: true
  }))));
}

var interval = vectorPlugin.createInternalPlugin(vectorPlugin._objectSpread2({
  component: IntervalComponent
}, props));

const Store = function Store() {
  const store = create__default['default'](() => ({
    data: {}
  }));
  this.useStore = store;
  const folders = {};
  const orderedPaths = new Set();

  this.getVisiblePaths = () => {
    const data = this.getData();
    const paths = Object.keys(data);
    const hiddenFolders = [];
    Object.entries(folders).forEach(([path, settings]) => {
      if (settings.render && paths.some(p => p.indexOf(path) === 0) && !settings.render(this.get)) hiddenFolders.push(path + '.');
    });
    const visiblePaths = [];
    orderedPaths.forEach(path => {
      if (path in data && data[path].__refCount > 0 && hiddenFolders.every(p => path.indexOf(p) === -1) && (!data[path].render || data[path].render(this.get))) visiblePaths.push(path);
    });
    return visiblePaths;
  };

  this.setOrderedPaths = newPaths => {
    newPaths.forEach(p => orderedPaths.add(p));
  };

  this.orderPaths = paths => {
    this.setOrderedPaths(paths);
    return paths;
  };

  this.disposePaths = paths => {
    store.setState(s => {
      const data = s.data;
      paths.forEach(path => {
        if (path in data) {
          const input = data[path];
          input.__refCount--;

          if (input.__refCount === 0 && input.type in vectorPlugin.SpecialInputTypes) {
            delete data[path];
          }
        }
      });
      return {
        data
      };
    });
  };

  this.dispose = () => {
    store.setState(() => {
      return {
        data: {}
      };
    });
  };

  this.getFolderSettings = path => {
    return folders[path] || {};
  };

  this.getData = () => {
    return store.getState().data;
  };

  this.addData = (newData, override) => {
    store.setState(s => {
      const data = s.data;
      Object.entries(newData).forEach(([path, newInputData]) => {
        let input = data[path];

        if (!!input) {
          const _ref = newInputData,
                {
            type,
            value
          } = _ref,
                rest = vectorPlugin._objectWithoutProperties(_ref, ["type", "value"]);

          if (type !== input.type) {
            vectorPlugin.warn(vectorPlugin.LevaErrors.INPUT_TYPE_OVERRIDE, type);
          } else {
            if (input.__refCount === 0 || override) {
              Object.assign(input, rest);
            }

            input.__refCount++;
          }
        } else {
          data[path] = vectorPlugin._objectSpread2(vectorPlugin._objectSpread2({}, newInputData), {}, {
            __refCount: 1
          });
        }
      });
      return {
        data
      };
    });
  };

  this.setValueAtPath = (path, value) => {
    store.setState(s => {
      const data = s.data;
      vectorPlugin.updateInput(data[path], value);
      return {
        data
      };
    });
  };

  this.setSettingsAtPath = (path, settings) => {
    store.setState(s => {
      const data = s.data;
      data[path].settings = vectorPlugin._objectSpread2(vectorPlugin._objectSpread2({}, data[path].settings), settings);
      return {
        data
      };
    });
  };

  this.disableInputAtPath = (path, flag) => {
    store.setState(s => {
      const data = s.data;
      data[path].disabled = flag;
      return {
        data
      };
    });
  };

  this.set = values => {
    store.setState(s => {
      const data = s.data;
      Object.entries(values).forEach(([path, value]) => {
        try {
          vectorPlugin.updateInput(data[path], value);
        } catch (_unused) {}
      });
      return {
        data
      };
    });
  };

  this.get = path => {
    try {
      return this.getData()[path].value;
    } catch (e) {
      vectorPlugin.warn(vectorPlugin.LevaErrors.PATH_DOESNT_EXIST, path);
    }
  };

  const _getDataFromSchema = (schema, rootPath, mappedPaths) => {
    const data = {};
    Object.entries(schema).forEach(([key, input]) => {
      let newPath = join(rootPath, key);

      if (input.type === vectorPlugin.SpecialInputTypes.FOLDER) {
        const newData = _getDataFromSchema(input.schema, newPath, mappedPaths);

        Object.assign(data, newData);
        if (!(newPath in folders)) folders[newPath] = input.settings;
      } else if (key in mappedPaths) {
        vectorPlugin.warn(vectorPlugin.LevaErrors.DUPLICATE_KEYS, key, newPath, mappedPaths[key]);
      } else {
        mappedPaths[key] = newPath;
        let _render = undefined;
        let _label = undefined;
        let _hint = undefined;

        let _optional;

        let _disabled;

        let _input = input;

        if (typeof input === 'object' && !Array.isArray(input)) {
          const {
            render,
            label,
            optional,
            disabled,
            hint
          } = input,
                rest = vectorPlugin._objectWithoutProperties(input, ["render", "label", "optional", "disabled", "hint"]);

          _label = label;
          _render = render;
          _input = rest;
          _optional = optional;
          _disabled = disabled;
          _hint = hint;
        }

        const normalizedInput = vectorPlugin.normalizeInput(_input, newPath);

        if (normalizedInput) {
          var _label2;

          data[newPath] = normalizedInput;
          data[newPath].key = key;
          data[newPath].label = (_label2 = _label) !== null && _label2 !== void 0 ? _label2 : key;
          data[newPath].hint = _hint;

          if (!(input.type in vectorPlugin.SpecialInputTypes)) {
            var _optional2, _disabled2;

            data[newPath].optional = (_optional2 = _optional) !== null && _optional2 !== void 0 ? _optional2 : false;
            data[newPath].disabled = (_disabled2 = _disabled) !== null && _disabled2 !== void 0 ? _disabled2 : false;
          }

          if (typeof _render === 'function') data[newPath].render = _render;
        }
      }
    });
    return data;
  };

  this.getDataFromSchema = schema => {
    const mappedPaths = {};

    const data = _getDataFromSchema(schema, '', mappedPaths);

    return [data, mappedPaths];
  };
};
const levaStore = new Store();
function useCreateStore() {
  return React.useMemo(() => new Store(), []);
}

if (process.env.NODE_ENV === 'development' && typeof window !== 'undefined') {
  window.__LEVA__STORE = levaStore;
}

const defaultSettings$1 = {
  collapsed: false
};
function folder(schema, settings) {
  return {
    type: vectorPlugin.SpecialInputTypes.FOLDER,
    schema,
    settings: vectorPlugin._objectSpread2(vectorPlugin._objectSpread2({}, defaultSettings$1), settings)
  };
}

function button(onClick) {
  return {
    type: vectorPlugin.SpecialInputTypes.BUTTON,
    onClick
  };
}

function buttonGroup(opts) {
  return {
    type: vectorPlugin.SpecialInputTypes.BUTTON_GROUP,
    opts
  };
}

const defaultSettings = {
  graph: false,
  interval: 100
};
function monitor(objectOrFn, settings) {
  return {
    type: vectorPlugin.SpecialInputTypes.MONITOR,
    objectOrFn,
    settings: vectorPlugin._objectSpread2(vectorPlugin._objectSpread2({}, defaultSettings), settings)
  };
}

const isInput = v => '__levaInput' in v;
const buildTree = (paths, filter) => {
  const tree = {};

  const _filter = filter ? filter.toLowerCase() : null;

  paths.forEach(path => {
    const [valueKey, folderPath] = getKeyPath(path);

    if (!_filter || valueKey.toLowerCase().indexOf(_filter) > -1) {
      merge__default['default'](tree, folderPath, {
        [valueKey]: {
          __levaInput: true,
          path
        }
      });
    }
  });
  return tree;
};

function FolderTitle({
  toggle,
  toggled,
  name
}) {
  return React__default['default'].createElement(vectorPlugin.StyledTitle, {
    onClick: () => toggle()
  }, React__default['default'].createElement(vectorPlugin.Chevron, {
    toggled: toggled
  }), React__default['default'].createElement("div", null, name));
}

function ControlInput(_ref) {
  let {
    type,
    label,
    path,
    valueKey,
    value,
    settings,
    setValue,
    disabled
  } = _ref,
      rest = vectorPlugin._objectWithoutProperties(_ref, ["type", "label", "path", "valueKey", "value", "settings", "setValue", "disabled"]);

  const {
    displayValue,
    onChange,
    onUpdate
  } = vectorPlugin.useValue({
    type,
    value,
    settings,
    setValue
  });
  const Input = vectorPlugin.Plugins[type].component;

  if (!Input) {
    vectorPlugin.log(vectorPlugin.LevaErrors.NO_COMPONENT_FOR_TYPE, type, path);
    return null;
  }

  return React__default['default'].createElement(vectorPlugin.InputContext.Provider, {
    value: vectorPlugin._objectSpread2({
      key: valueKey,
      path,
      id: 'leva__' + path,
      label,
      displayValue,
      value,
      onChange,
      onUpdate,
      settings,
      setValue,
      disabled
    }, rest)
  }, React__default['default'].createElement(vectorPlugin.StyledInputWrapper, {
    disabled: disabled
  }, React__default['default'].createElement(Input, null)));
}

const StyledButton = vectorPlugin.styled('button', {
  display: 'block',
  $leva__reset: '',
  fontWeight: '$leva__button',
  color: '$leva__highlight3',
  height: '$leva__rowHeight',
  borderStyle: 'none',
  borderRadius: '$leva__sm',
  backgroundColor: '$leva__accent2',
  cursor: 'pointer',
  $leva__hover: '$leva__accent3',
  $leva__active: '$leva__accent3 $leva__accent1',
  $leva__focus: ''
});

function Button({
  onClick,
  label
}) {
  return React__default['default'].createElement(vectorPlugin.Row, null, React__default['default'].createElement(StyledButton, {
    onClick: () => onClick()
  }, label));
}

const StyledButtonGroup = vectorPlugin.styled('div', {
  $leva__flex: '',
  justifyContent: 'flex-end',
  gap: '$leva__colGap'
});

const StyledButtonGroupButton = vectorPlugin.styled('button', {
  $leva__reset: '',
  cursor: 'pointer',
  borderRadius: '$leva__xs',
  '&:hover': {
    backgroundColor: '$leva__elevation3'
  }
});

function ButtonGroup({
  label,
  opts
}) {
  return React__default['default'].createElement(vectorPlugin.Row, {
    input: true
  }, React__default['default'].createElement(vectorPlugin.Label, null, label), React__default['default'].createElement(StyledButtonGroup, null, Object.entries(opts).map(([label, onClick]) => React__default['default'].createElement(StyledButtonGroupButton, {
    key: label,
    onClick: () => onClick()
  }, label))));
}

const Canvas = vectorPlugin.styled('canvas', {
  height: '$leva__monitorHeight',
  width: '100%',
  display: 'block',
  borderRadius: '$leva__sm'
});

const POINTS = 100;

function push(arr, val) {
  arr.push(val);
  if (arr.length > POINTS) arr.shift();
}

const MonitorCanvas = React.forwardRef(function ({
  initialValue
}, ref) {
  const accentColor = vectorPlugin.useTh('colors', 'leva__highlight3');
  const backgroundColor = vectorPlugin.useTh('colors', 'leva__elevation2');
  const fillColor = vectorPlugin.useTh('colors', 'leva__highlight1');
  const [gradientTop, gradientBottom] = React.useMemo(() => {
    return [tc__default['default'](fillColor).setAlpha(0.4).toRgbString(), tc__default['default'](fillColor).setAlpha(0.1).toRgbString()];
  }, [fillColor]);
  const points = React.useRef([initialValue]);
  const min = React.useRef(initialValue);
  const max = React.useRef(initialValue);
  const raf = React.useRef();
  const drawPlot = React.useCallback((_canvas, _ctx) => {
    if (!_canvas) return;
    const {
      width,
      height
    } = _canvas;
    const path = new Path2D();
    const interval = width / POINTS;
    const verticalPadding = height * 0.05;

    for (let i = 0; i < points.current.length; i++) {
      const p = vectorPlugin.range(points.current[i], min.current, max.current);
      const x = interval * i;
      const y = height - p * (height - verticalPadding * 2) - verticalPadding;
      path.lineTo(x, y);
    }

    _ctx.clearRect(0, 0, width, height);

    const gradientPath = new Path2D(path);
    gradientPath.lineTo(interval * (points.current.length + 1), height);
    gradientPath.lineTo(0, height);
    gradientPath.lineTo(0, 0);

    const gradient = _ctx.createLinearGradient(0, 0, 0, height);

    gradient.addColorStop(0.0, gradientTop);
    gradient.addColorStop(1.0, gradientBottom);
    _ctx.fillStyle = gradient;

    _ctx.fill(gradientPath);

    _ctx.strokeStyle = backgroundColor;
    _ctx.lineJoin = 'round';
    _ctx.lineWidth = 14;

    _ctx.stroke(path);

    _ctx.strokeStyle = accentColor;
    _ctx.lineWidth = 2;

    _ctx.stroke(path);
  }, [accentColor, backgroundColor, gradientTop, gradientBottom]);
  const [canvas, ctx] = vectorPlugin.useCanvas2d(drawPlot);
  React.useImperativeHandle(ref, () => ({
    frame: val => {
      if (min.current === undefined || val < min.current) min.current = val;
      if (max.current === undefined || val > max.current) max.current = val;
      push(points.current, val);
      raf.current = requestAnimationFrame(() => drawPlot(canvas.current, ctx.current));
    }
  }), [canvas, ctx, drawPlot]);
  React.useEffect(() => () => cancelAnimationFrame(raf.current), []);
  return React__default['default'].createElement(Canvas, {
    ref: canvas
  });
});

const parse = val => Number.isFinite(val) ? val.toPrecision(2) : val.toString();

const MonitorLog = React.forwardRef(function ({
  initialValue
}, ref) {
  const [val, set] = React.useState(parse(initialValue));
  React.useImperativeHandle(ref, () => ({
    frame: v => set(parse(v))
  }), []);
  return React__default['default'].createElement("div", null, val);
});

function getValue(o) {
  return typeof o === 'function' ? o() : o.current;
}

function Monitor({
  label,
  objectOrFn,
  settings
}) {
  const ref = React.useRef();
  const initialValue = React.useRef(getValue(objectOrFn));
  React.useEffect(() => {
    const timeout = window.setInterval(() => {
      var _ref$current;

      return (_ref$current = ref.current) === null || _ref$current === void 0 ? void 0 : _ref$current.frame(getValue(objectOrFn));
    }, settings.interval);
    return () => window.clearInterval(timeout);
  }, [objectOrFn, settings.interval]);
  return React__default['default'].createElement(vectorPlugin.Row, {
    input: true
  }, React__default['default'].createElement(vectorPlugin.Label, {
    align: "top"
  }, label), settings.graph ? React__default['default'].createElement(MonitorCanvas, {
    ref: ref,
    initialValue: initialValue.current
  }) : React__default['default'].createElement(MonitorLog, {
    ref: ref,
    initialValue: initialValue.current
  }));
}

const specialComponents = {
  [vectorPlugin.SpecialInputTypes.BUTTON]: Button,
  [vectorPlugin.SpecialInputTypes.BUTTON_GROUP]: ButtonGroup,
  [vectorPlugin.SpecialInputTypes.MONITOR]: Monitor
};
const Control = React__default['default'].memo(({
  path
}) => {
  const [input, {
    set,
    setSettings,
    disable
  }] = vectorPlugin.useInput(path);
  if (!input) return null;

  const {
    type,
    label,
    key
  } = input,
        inputProps = vectorPlugin._objectWithoutProperties(input, ["type", "label", "key"]);

  if (type in vectorPlugin.SpecialInputTypes) {
    const SpecialInputForType = specialComponents[type];
    return React__default['default'].createElement(SpecialInputForType, vectorPlugin._extends({
      label: label,
      path: path
    }, inputProps));
  }

  if (!(type in vectorPlugin.Plugins)) {
    vectorPlugin.log(vectorPlugin.LevaErrors.UNSUPPORTED_INPUT, type, path);
    return null;
  }

  return React__default['default'].createElement(ControlInput, vectorPlugin._extends({
    type: type,
    label: label,
    path: path,
    valueKey: key,
    setValue: set,
    setSettings: setSettings,
    disable: disable
  }, inputProps));
});

const Folder = ({
  name,
  path,
  tree
}) => {
  const store = vectorPlugin.useStoreContext();
  const newPath = join(path, name);
  const {
    collapsed
  } = store.getFolderSettings(newPath);
  const [toggled, setToggle] = React.useState(!collapsed);
  return React__default['default'].createElement(vectorPlugin.StyledFolder, null, React__default['default'].createElement(FolderTitle, {
    name: name,
    toggled: toggled,
    toggle: () => setToggle(t => !t)
  }), React__default['default'].createElement(TreeWrapper, {
    parent: newPath,
    tree: tree,
    toggled: toggled
  }));
};

const TreeWrapper = React__default['default'].memo(({
  isRoot: _isRoot = false,
  fill: _fill = false,
  flat: _flat = false,
  parent,
  tree,
  toggled
}) => {
  const {
    wrapperRef,
    contentRef
  } = useToggle(toggled);
  return React__default['default'].createElement(vectorPlugin.StyledWrapper, {
    ref: wrapperRef,
    isRoot: _isRoot,
    fill: _fill,
    flat: _flat
  }, React__default['default'].createElement(vectorPlugin.StyledContent, {
    ref: contentRef,
    isRoot: _isRoot,
    toggled: toggled
  }, Object.entries(tree).map(([key, value]) => isInput(value) ? React__default['default'].createElement(Control, {
    key: value.path,
    valueKey: value.valueKey,
    path: value.path
  }) : React__default['default'].createElement(Folder, {
    key: key,
    name: key,
    path: parent,
    tree: value
  }))));
});

const StyledRoot = vectorPlugin.styled('div', {
  position: 'relative',
  fontFamily: '$leva__mono',
  fontSize: '$leva__root',
  color: '$leva__rootText',
  backgroundColor: '$leva__elevation1',
  variants: {
    fill: {
      false: {
        position: 'fixed',
        top: '10px',
        right: '10px',
        zIndex: 1000,
        width: '$leva__rootWidth'
      },
      true: {
        position: 'relative',
        width: '100%'
      }
    },
    flat: {
      false: {
        borderRadius: '$leva__lg',
        boxShadow: '$leva__level1'
      }
    },
    oneLineLabels: {
      true: {
        [`${vectorPlugin.StyledInputRow}`]: {
          gridTemplateColumns: 'auto',
          gridAutoColumns: 'minmax(max-content, 1fr)',
          gridAutoRows: 'minmax($sizes$leva__rowHeight), auto)',
          rowGap: 0,
          columnGap: 0,
          marginTop: '$leva__rowGap'
        }
      }
    },
    hideTitleBar: {
      true: {
        $$titleBarHeight: '0px'
      },
      false: {
        $$titleBarHeight: '$sizes$leva__titleBarHeight'
      }
    }
  },
  '&,*,*:after,*:before': {
    boxSizing: 'border-box'
  },
  '*::selection': {
    backgroundColor: '$leva__accent2'
  }
});

const Icon = vectorPlugin.styled('i', {
  $leva__flexCenter: '',
  width: 40,
  WebkitUserSelect: 'none',
  userSelect: 'none',
  cursor: 'pointer',
  '> svg': {
    fill: '$leva__highlight1',
    transition: 'transform 350ms ease, fill 250ms ease'
  },
  '&:hover > svg': {
    fill: '$leva__highlight3'
  },
  variants: {
    active: {
      true: {
        '> svg': {
          fill: '$leva__highlight2'
        }
      }
    }
  }
});
const StyledTitleWithFilter = vectorPlugin.styled('div', {
  display: 'flex',
  alignItems: 'stretch',
  justifyContent: 'space-between',
  height: '$leva__titleBarHeight',
  cursor: 'grab'
});
const FilterWrapper = vectorPlugin.styled('div', {
  $leva__flex: '',
  position: 'relative',
  width: '100%',
  overflow: 'hidden',
  transition: 'height 250ms ease',
  color: '$leva__highlight3',
  paddingLeft: '$leva__md',
  [`> ${Icon}`]: {
    height: 30
  },
  variants: {
    toggled: {
      true: {
        height: 30
      },
      false: {
        height: 0
      }
    }
  }
});
const StyledFilterInput = vectorPlugin.styled('input', {
  $leva__reset: '',
  flex: 1,
  position: 'relative',
  height: 30,
  width: '100%',
  backgroundColor: 'transparent',
  fontSize: '10px',
  borderRadius: '$leva__root',
  '&:focus': {},
  '&::placeholder': {
    color: '$leva__highlight2'
  }
});
const Drag = vectorPlugin.styled('div', {
  $leva__flexCenter: '',
  $leva__draggable: '',
  flex: 1,
  '> svg': {
    fill: '$leva__highlight1',
    transition: 'fill 250ms ease'
  },
  '&:hover > svg': {
    fill: '$leva__highlight3'
  }
});

const FilterInput = React__default['default'].forwardRef(({
  setFilter
}, ref) => {
  const [value, set] = React.useState('');
  const debouncedOnChange = React.useMemo(() => vectorPlugin.debounce(setFilter, 250), [setFilter]);

  const clear = () => {
    setFilter('');
    set('');
  };

  const _onChange = e => {
    const v = e.currentTarget.value;
    set(v);
  };

  React.useEffect(() => {
    debouncedOnChange(value);
  }, [value, debouncedOnChange]);
  return React__default['default'].createElement(React__default['default'].Fragment, null, React__default['default'].createElement(StyledFilterInput, {
    ref: ref,
    value: value,
    placeholder: "[Open filter with CMD+SHIFT+L]",
    onPointerDown: e => e.stopPropagation(),
    onChange: _onChange
  }), React__default['default'].createElement(Icon, {
    onClick: () => clear(),
    style: {
      visibility: value ? 'visible' : 'hidden'
    }
  }, React__default['default'].createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    height: "14",
    width: "14",
    viewBox: "0 0 20 20",
    fill: "currentColor"
  }, React__default['default'].createElement("path", {
    fillRule: "evenodd",
    d: "M10 18a8 8 0 100-16 8 8 0 000 16zM8.707 7.293a1 1 0 00-1.414 1.414L8.586 10l-1.293 1.293a1 1 0 101.414 1.414L10 11.414l1.293 1.293a1 1 0 001.414-1.414L11.414 10l1.293-1.293a1 1 0 00-1.414-1.414L10 8.586 8.707 7.293z",
    clipRule: "evenodd"
  }))));
});
function TitleWithFilter({
  setFilter,
  onDrag,
  toggle,
  toggled
}) {
  const [filterShown, setShowFilter] = React.useState(false);
  const inputRef = React.useRef(null);
  React.useEffect(() => {
    var _inputRef$current, _inputRef$current2;

    if (filterShown) (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.focus();else (_inputRef$current2 = inputRef.current) === null || _inputRef$current2 === void 0 ? void 0 : _inputRef$current2.blur();
  }, [filterShown]);
  const bind = reactUseGesture.useDrag(({
    offset: [x, y]
  }) => onDrag({
    x,
    y
  }), {
    filterTaps: true
  });
  React.useEffect(() => {
    const handleShortcut = event => {
      if (event.key === 'L' && event.shiftKey && event.metaKey) {
        setShowFilter(f => !f);
      }
    };

    window.addEventListener('keydown', handleShortcut);
    return () => window.removeEventListener('keydown', handleShortcut);
  }, []);
  return React__default['default'].createElement(React__default['default'].Fragment, null, React__default['default'].createElement(StyledTitleWithFilter, null, React__default['default'].createElement(Icon, {
    active: !toggled,
    onClick: () => toggle()
  }, React__default['default'].createElement(vectorPlugin.Chevron, {
    toggled: toggled,
    width: 12,
    height: 8
  })), React__default['default'].createElement(Drag, bind(), React__default['default'].createElement("svg", {
    width: "20",
    height: "10",
    viewBox: "0 0 28 14",
    xmlns: "http://www.w3.org/2000/svg"
  }, React__default['default'].createElement("circle", {
    cx: "2",
    cy: "2",
    r: "2"
  }), React__default['default'].createElement("circle", {
    cx: "14",
    cy: "2",
    r: "2"
  }), React__default['default'].createElement("circle", {
    cx: "26",
    cy: "2",
    r: "2"
  }), React__default['default'].createElement("circle", {
    cx: "2",
    cy: "12",
    r: "2"
  }), React__default['default'].createElement("circle", {
    cx: "14",
    cy: "12",
    r: "2"
  }), React__default['default'].createElement("circle", {
    cx: "26",
    cy: "12",
    r: "2"
  }))), React__default['default'].createElement(Icon, {
    active: filterShown,
    onClick: () => setShowFilter(f => !f)
  }, React__default['default'].createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    height: "20",
    viewBox: "0 0 20 20"
  }, React__default['default'].createElement("path", {
    d: "M9 9a2 2 0 114 0 2 2 0 01-4 0z"
  }), React__default['default'].createElement("path", {
    fillRule: "evenodd",
    d: "M10 18a8 8 0 100-16 8 8 0 000 16zm1-13a4 4 0 00-3.446 6.032l-2.261 2.26a1 1 0 101.414 1.415l2.261-2.261A4 4 0 1011 5z",
    clipRule: "evenodd"
  })))), React__default['default'].createElement(FilterWrapper, {
    toggled: filterShown
  }, React__default['default'].createElement(FilterInput, {
    ref: inputRef,
    setFilter: setFilter
  })));
}

function LevaRoot(_ref) {
  let {
    store,
    hidden = false,
    theme,
    collapsed = false
  } = _ref,
      props = vectorPlugin._objectWithoutProperties(_ref, ["store", "hidden", "theme", "collapsed"]);

  const themeContext = useDeepMemo(() => vectorPlugin.mergeTheme(theme), [theme]);
  const [toggled, setToggle] = React.useState(!collapsed);
  if (!store || hidden) return null;
  return React__default['default'].createElement(vectorPlugin.ThemeContext.Provider, {
    value: themeContext
  }, React__default['default'].createElement(LevaCore, vectorPlugin._extends({
    store: store
  }, props, {
    toggled: toggled,
    setToggle: setToggle,
    rootClass: themeContext.className
  })));
}
const LevaCore = React__default['default'].memo(({
  store,
  rootClass,
  fill: _fill = false,
  flat: _flat = false,
  oneLineLabels: _oneLineLabels = false,
  hideTitleBar: _hideTitleBar = false,
  hideCopyButton: _hideCopyButton = false,
  toggled,
  setToggle
}) => {
  const paths = useVisiblePaths(store);
  const [filter, setFilter] = React.useState('');
  const tree = React.useMemo(() => buildTree(paths, filter), [paths, filter]);
  const [rootRef, set] = vectorPlugin.useTransform();
  const shouldShow = paths.length > 0;
  return React__default['default'].createElement(vectorPlugin.PanelSettingsContext.Provider, {
    value: {
      hideCopyButton: _hideCopyButton
    }
  }, React__default['default'].createElement(StyledRoot, {
    ref: rootRef,
    className: rootClass,
    fill: _fill,
    flat: _flat,
    oneLineLabels: _oneLineLabels,
    hideTitleBar: _hideTitleBar,
    style: {
      display: shouldShow ? 'block' : 'none'
    }
  }, !_hideTitleBar && React__default['default'].createElement(TitleWithFilter, {
    onDrag: set,
    setFilter: setFilter,
    toggle: () => setToggle(t => !t),
    toggled: toggled
  }), shouldShow && React__default['default'].createElement(vectorPlugin.StoreContext.Provider, {
    value: store
  }, React__default['default'].createElement(TreeWrapper, {
    isRoot: true,
    fill: _fill,
    flat: _flat,
    tree: tree,
    toggled: toggled
  }))));
});

let rootInitialized = false;
let rootEl = null;
function Leva(_ref) {
  let {
    isRoot = false
  } = _ref,
      props = vectorPlugin._objectWithoutProperties(_ref, ["isRoot"]);

  React.useEffect(() => {
    rootInitialized = true;

    if (!isRoot && rootEl) {
      rootEl.remove();
      rootEl = null;
    }

    return () => {
      if (!isRoot) rootInitialized = false;
    };
  }, [isRoot]);
  return React__default['default'].createElement(LevaRoot, vectorPlugin._extends({
    store: levaStore
  }, props));
}
function useRenderRoot(isGlobalPanel) {
  React.useEffect(() => {
    if (isGlobalPanel && !rootInitialized) {
      if (!rootEl) {
        rootEl = document.getElementById('leva__root') || Object.assign(document.createElement('div'), {
          id: 'leva__root'
        });

        if (document.body) {
          document.body.appendChild(rootEl);
          ReactDOM__default['default'].render(React__default['default'].createElement(Leva, {
            isRoot: true
          }), rootEl);
        }
      }

      rootInitialized = true;
    }
  }, [isGlobalPanel]);
}

function LevaPanel(_ref) {
  let {
    store
  } = _ref,
      props = vectorPlugin._objectWithoutProperties(_ref, ["store"]);

  const parentStore = vectorPlugin.useStoreContext();

  const _store = store === undefined ? parentStore : store;

  return React__default['default'].createElement(LevaRoot, vectorPlugin._extends({
    store: _store
  }, props));
}

function parseArgs(schemaOrFolderName, settingsOrDepsOrSchema, depsOrSettingsOrFolderSettings, depsOrSettings, depsOrUndefined) {
  let schema;
  let folderName = undefined;
  let folderSettings;
  let hookSettings;
  let deps;

  if (typeof schemaOrFolderName === 'string') {
    folderName = schemaOrFolderName;
    schema = settingsOrDepsOrSchema;

    if (Array.isArray(depsOrSettingsOrFolderSettings)) {
      deps = depsOrSettingsOrFolderSettings;
    } else {
      if (depsOrSettingsOrFolderSettings) {
        if ('store' in depsOrSettingsOrFolderSettings) {
          hookSettings = depsOrSettingsOrFolderSettings;
          deps = depsOrSettings;
        } else {
          folderSettings = depsOrSettingsOrFolderSettings;

          if (Array.isArray(depsOrSettings)) {
            deps = depsOrSettings;
          } else {
            hookSettings = depsOrSettings;
            deps = depsOrUndefined;
          }
        }
      }
    }
  } else {
    schema = schemaOrFolderName;

    if (Array.isArray(settingsOrDepsOrSchema)) {
      deps = settingsOrDepsOrSchema;
    } else {
      hookSettings = settingsOrDepsOrSchema;
      deps = depsOrSettingsOrFolderSettings;
    }
  }

  return {
    schema,
    folderName,
    folderSettings,
    hookSettings,
    deps: deps || []
  };
}

function useControls(schemaOrFolderName, settingsOrDepsOrSchema, depsOrSettingsOrFolderSettings, depsOrSettings, depsOrUndefined) {
  const {
    folderName,
    schema,
    folderSettings,
    hookSettings,
    deps
  } = parseArgs(schemaOrFolderName, settingsOrDepsOrSchema, depsOrSettingsOrFolderSettings, depsOrSettings, depsOrUndefined);
  const schemaIsFunction = typeof schema === 'function';
  const depsChanged = React.useRef(false);
  const firstRender = React.useRef(true);

  const _schema = useDeepMemo(() => {
    depsChanged.current = true;
    const s = typeof schema === 'function' ? schema() : schema;
    return folderName ? {
      [folderName]: folder(s, folderSettings)
    } : s;
  }, deps);

  const isGlobalPanel = !(hookSettings !== null && hookSettings !== void 0 && hookSettings.store);
  useRenderRoot(isGlobalPanel);
  const [store] = React.useState(() => (hookSettings === null || hookSettings === void 0 ? void 0 : hookSettings.store) || levaStore);
  const [initialData, mappedPaths] = React.useMemo(() => store.getDataFromSchema(_schema), [store, _schema]);
  const paths = React.useMemo(() => store.orderPaths(Object.values(mappedPaths)), [mappedPaths, store]);
  const values = useValuesForPath(store, paths, initialData);
  const set = React.useCallback(values => {
    const _values = Object.entries(values).reduce((acc, [p, v]) => Object.assign(acc, {
      [mappedPaths[p]]: v
    }), {});

    store.set(_values);
  }, [store, mappedPaths]);
  React.useEffect(() => {
    const shouldOverrideSettings = !firstRender.current && depsChanged.current;
    store.addData(initialData, shouldOverrideSettings);
    firstRender.current = false;
    depsChanged.current = false;
    return () => store.disposePaths(paths);
  }, [store, paths, initialData]);
  if (schemaIsFunction) return [values, set];
  return values;
}

vectorPlugin.register('SELECT', select);
vectorPlugin.register('IMAGE', image);
vectorPlugin.register('NUMBER', vectorPlugin.number);
vectorPlugin.register('COLOR', color);
vectorPlugin.register('STRING', string);
vectorPlugin.register('BOOLEAN', boolean);
vectorPlugin.register('INTERVAL', interval);
vectorPlugin.register('VECTOR3D', vector3d);
vectorPlugin.register('VECTOR2D', vector2d);

exports.LevaStoreProvider = vectorPlugin.LevaStoreProvider;
exports.useStoreContext = vectorPlugin.useStoreContext;
exports.Leva = Leva;
exports.LevaPanel = LevaPanel;
exports.button = button;
exports.buttonGroup = buttonGroup;
exports.folder = folder;
exports.levaStore = levaStore;
exports.monitor = monitor;
exports.useControls = useControls;
exports.useCreateStore = useCreateStore;
