// leva/plugin/dist/leva-plugin.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tc = require('tinycolor2');
var vectorPlugin = require('../../dist/vector-plugin-19a8a46a.cjs.dev.js');
require('dequal/lite');
require('react');
require('zustand/shallow');
require('@radix-ui/react-portal');
require('@stitches/react');
require('react-use-gesture');
require('@radix-ui/react-tooltip');
require('clipboard-polyfill/text');
require('v8n');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var tc__default = /*#__PURE__*/_interopDefault(tc);



Object.defineProperty(exports, 'tinycolor2', {
	enumerable: true,
	get: function () {
		return tc__default['default'];
	}
});
exports.Label = vectorPlugin.Label;
exports.Portal = vectorPlugin.Portal;
exports.Row = vectorPlugin.Row;
Object.defineProperty(exports, 'SpecialInputTypes', {
	enumerable: true,
	get: function () {
		return vectorPlugin.SpecialInputTypes;
	}
});
exports.ValueInput = vectorPlugin.ValueInput;
exports.Vector = vectorPlugin.Vector;
exports.clamp = vectorPlugin.clamp;
exports.createPlugin = vectorPlugin.createPlugin;
exports.debounce = vectorPlugin.debounce;
exports.evaluate = vectorPlugin.evaluate;
exports.formatVector = vectorPlugin.formatVector;
exports.getVectorPlugin = vectorPlugin.getVectorPlugin;
exports.getVectorSchema = vectorPlugin.getVectorSchema;
exports.getVectorType = vectorPlugin.getVectorType;
exports.normalizeKeyedNumberSettings = vectorPlugin.normalizeKeyedNumberSettings;
exports.normalizeVector = vectorPlugin.normalizeVector;
exports.pad = vectorPlugin.pad;
exports.sanitizeVector = vectorPlugin.sanitizeVector;
exports.styled = vectorPlugin.styled;
exports.useCanvas2d = vectorPlugin.useCanvas2d;
exports.useDrag = vectorPlugin.useDrag;
exports.useDragNumber = vectorPlugin.useDragNumber;
exports.useInput = vectorPlugin.useInput;
exports.useInputContext = vectorPlugin.useInputContext;
exports.useTh = vectorPlugin.useTh;
exports.useTransform = vectorPlugin.useTransform;
