// @three-material-editor/react/dist/react.cjs.development.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var reactThreeFiber = require('react-three-fiber');
var core = require('@three-material-editor/core');
var reactDom = require('react-dom');

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;

  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }

  return target;
}

var Html = /*#__PURE__*/React.forwardRef(function (_ref, ref) {
  var _portal$current;

  var children = _ref.children,
      className = _ref.className,
      portal = _ref.portal,
      props = _objectWithoutPropertiesLoose(_ref, ["children", "style", "className", "portal"]);

  // const { gl } = useThree();
  var _useState = React.useState(function () {
    return document.createElement('div');
  }),
      el = _useState[0];

  var group = React.useRef(null);
  var target = (_portal$current = portal == null ? void 0 : portal.current) != null ? _portal$current : document.body;
  React.useEffect(function () {
    if (group.current) {
      if (target) {
        target.appendChild(el);
      }

      return function () {
        if (target) target.removeChild(el);
        reactDom.unmountComponentAtNode(el);
      };
    }

    return undefined;
  }, [target, el]);
  React.useEffect(function () {
    return void reactDom.render( /*#__PURE__*/React__default.createElement("div", {
      ref: ref,
      className: className,
      children: children
    }), el);
  });
  return /*#__PURE__*/React__default.createElement("group", Object.assign({}, props, {
    ref: group
  }));
});

var optionsDefault = {
  overrideRaf: false,
  fullScreen: true
};
exports.Logic = function Logic() {
  return null;
};
exports.MaterialEditor = function MaterialEditor(_options) {};
exports.useEditorComposer = function useEditorComposer() {};

{
  exports.Logic = function Logic() {
    var _useThree = reactThreeFiber.useThree(),
        scene = _useThree.scene,
        gl = _useThree.gl;

    reactThreeFiber.useFrame(function () {
      core.materialsToProgram(scene, gl);
    });
    return null;
  };

  exports.useEditorComposer = function useEditorComposer() {
    var onRefChange = React.useCallback(function (node) {
      if (node === null) {
        core.editorContext.composer = null;
      } else {
        core.editorContext.composer = node;
      }
    }, []);
    return onRefChange;
  };

  exports.MaterialEditor = function MaterialEditor(_options) {
    var options = Object.assign(optionsDefault, _options);
    Object.assign(core.editorState, options);
    return /*#__PURE__*/React__default.createElement(React__default.Fragment, null, /*#__PURE__*/React__default.createElement(exports.Logic, null), /*#__PURE__*/React__default.createElement(Html, null, /*#__PURE__*/React__default.createElement(core.EditorDom, {
      className: core.editorState.className
    })));
  };
}
